// @graphql-mesh/utils/cjs/wrapFetchWithHooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapFetchWithHooks = wrapFetchWithHooks;
const iterateAsync_js_1 = require("./iterateAsync.js");
const map_maybe_promise_js_1 = require("./map-maybe-promise.js");
function wrapFetchWithHooks(onFetchHooks) {
    return function wrappedFetchFn(url, options, context, info) {
        let fetchFn;
        const onFetchDoneHooks = [];
        return (0, map_maybe_promise_js_1.mapMaybePromise)((0, iterateAsync_js_1.iterateAsync)(onFetchHooks, onFetch => onFetch({
            fetchFn,
            setFetchFn(newFetchFn) {
                fetchFn = newFetchFn;
            },
            url,
            setURL(newUrl) {
                url = String(newUrl);
            },
            options,
            setOptions(newOptions) {
                options = newOptions;
            },
            context,
            info,
        }), onFetchDoneHooks), function handleIterationResult() {
            const res$ = fetchFn(url, options, context, info);
            if (onFetchDoneHooks.length === 0) {
                return res$;
            }
            return (0, map_maybe_promise_js_1.mapMaybePromise)(res$, function (response) {
                return (0, map_maybe_promise_js_1.mapMaybePromise)((0, iterateAsync_js_1.iterateAsync)(onFetchDoneHooks, onFetchDone => onFetchDone({
                    response,
                    setResponse(newResponse) {
                        response = newResponse;
                    },
                })), function handleOnFetchDone() {
                    return response;
                });
            });
        });
    };
}
